unit-testing-best-practices.html

```diff
@@ -0,0 +1,179 @@
+    </blockquote>
+
+
+    <p>Over the coming series of blog posts I will attempt to cover the questions above when it comes to unit
+      testing in Xcode. In this article I’ll broach some overarching concepts mostly taken from Robert Martin’s
+      excellent book <cite>Clean Code</cite>.</p>
+    <h2>F.I.R.S.T</h2>
+    <p>All tests should follow the FIRST principles</p>
+    <h3>Fast</h3>
+    <p>We want to run these tests regularly and often by smashing command U whenever we’ve made changes or we're
+      about to commit our work. For this to work we need our tests to be fast. How fast is fast? We’re going to be
+      writing a lot of tests so at the very minimum a test should be under 0.1 seconds.</p>
+    <h3>Independent</h3>
+    <p>Don’t write tests that depend on each other.
+      Xcode actively helps us with this, every function in an XCTestCase class that begins with “test” will receive a
+      new instance of your test class so no variables can be carried over.
+      <code>setup()</code> and <code>teardown()</code> can be used to clear persistent resources but it would be better
+      still to mock them. You should be able to run your tests in any order and have them pass.</p>
+
+    <h3>Repeatable</h3>
+    <p>Tests should run on any environment. If your unit tests only work on your laptop, your companies VPN or with a
+      network connection they can’t be run reliably and your team will second guess why they’ve failed or ignore them.</p>
+
+    <h3>Self-Validating</h3>
+    <p>Tests should simply fail or pass and not rely on a log.
+      <br>
+      I slipped up on this point when writing one of my first “unit tests” which tested the parsing of phone numbers from
+      email text. It had a giant log with the % of phone numbers we parsed alongside a print out of all the ones which had
+      failed and a spreadsheet I’d keep up to date the results for each version of our app.</p>
+
+    <h3>Timely</h3>
+    <p>Write them with, or better yet before your code. If you do it after your production code may be difficult to test.</p>
+
+    <h2>Clean Tests</h2>
+
+    <p>Do not adopt a dual standard where readability is neglected for tests. As your code changes your tests will have to
+      change people who did not initially write them will have to understand them so that they can adapt them, extend them
+      or remove tests that are no longer relevant without losing coverage. XCTest classes can and should have instance
+      variables and helper functions to enhance readability and reduce repetition.</p>
+
+    <h2>Smaller Tests</h2>
+    <p>Minimize the number of asserts and test just one concept per test function. Smaller tests are easier to read and
+      give clearer failure results. Any repetitive setup involved in doing this can be handled by <code>setup()</code> or
+      private helper functions</p>
+
+    <figure>
+      <pre>
+      <code>
+<span class="keyword">func</span> testAddMonths() {
```

Comment:
Indentation a bit off.

Reply:
The weird thing is when I try and correct this it actually alters the code. I'm using a third party tool to generate the snippet HTML although it appears anything inside `<pre>` the indentation actually impacts the appearance of the rendered html 😨 

Bit annoying as it looks pretty bad.

Reply:
Ah indeed, it does affect the layout.